### PR TITLE
Fix C++ parser segfaults when the parent node of a variable

### LINKED ISSUE #415: C++ parser segfaults on PCL
I have tried to parse the Point Cloud Library (PCL) with CodeCompass.
[PCL](https://github.com/PointCloudLibrary/pcl) is a popular library written in C++ with a CMake build system and shall be "compatible" with CodeCompass.

I have built PCL (with GCC though), but then the C++ parser fails on the first TU with a segmentation fault (running on 1 cores):
```
[INFO] [cppparser] parse started!
[INFO] (1/389) Parsing /projects/pcl/common/src/point_types.cpp
warning: unknown warning option '-Wabi=11' [-Wunknown-warning-option]
Segmentation fault (core dumped)
```

Running with Valgrind:
```
[INFO] [cppparser] parse started!
[INFO] (1/389) Parsing /projects/pcl/common/src/point_types.cpp
warning: unknown warning option '-Wabi=11' [-Wunknown-warning-option]
==12408== Invalid read of size 1
==12408==    at 0xA77F860: cc::parser::getMangledName[abi:cxx11](clang::MangleContext*, clang::NamedDecl const*, cc::model::FileLoc const&) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA77FC9B: cc::parser::getMangledName[abi:cxx11](clang::MangleContext*, clang::NamedDecl const*, cc::model::FileLoc const&) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA738FC2: cc::parser::ClangASTVisitor::VisitVarDecl(clang::VarDecl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73B73C: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseDecl(clang::Decl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73C471: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseDeclStmt(clang::DeclStmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA719913: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseStmt(clang::Stmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) [clone .constprop.7422] (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73AD40: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseDecl(clang::Decl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73F3F8: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseCapturedStmt(clang::CapturedStmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA719913: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseStmt(clang::Stmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) [clone .constprop.7422] (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA74D426: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseFunctionHelper(clang::FunctionDecl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73BBA0: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseDecl(clang::Decl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73C3ED: cc::parser::ClangASTVisitor::TraverseDecl(clang::Decl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==  Address 0x1c is not stack'd, malloc'd or (recently) free'd
==12408==
==12408==
==12408== Process terminating with default action of signal 11 (SIGSEGV): dumping core
==12408==  Access not within mapped region at address 0x1C
==12408==    at 0xA77F860: cc::parser::getMangledName[abi:cxx11](clang::MangleContext*, clang::NamedDecl const*, cc::model::FileLoc const&) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA77FC9B: cc::parser::getMangledName[abi:cxx11](clang::MangleContext*, clang::NamedDecl const*, cc::model::FileLoc const&) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA738FC2: cc::parser::ClangASTVisitor::VisitVarDecl(clang::VarDecl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73B73C: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseDecl(clang::Decl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73C471: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseDeclStmt(clang::DeclStmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA719913: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseStmt(clang::Stmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) [clone .constprop.7422] (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73AD40: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseDecl(clang::Decl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73F3F8: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseCapturedStmt(clang::CapturedStmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA719913: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseStmt(clang::Stmt*, llvm::SmallVectorImpl<llvm::PointerIntPair<clang::Stmt*, 1u, bool, llvm::PointerLikeTypeTraits<clang::Stmt*>, llvm::PointerIntPairInfo<clang::Stmt*, 1u, llvm::PointerLikeTypeTraits<clang::Stmt*> > > >*) [clone .constprop.7422] (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA74D426: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseFunctionHelper(clang::FunctionDecl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73BBA0: clang::RecursiveASTVisitor<cc::parser::ClangASTVisitor>::TraverseDecl(clang::Decl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==    by 0xA73C3ED: cc::parser::ClangASTVisitor::TraverseDecl(clang::Decl*) (in /codecompass/lib/parserplugin/libcppparser.so)
==12408==  If you believe this happened as a result of a stack
==12408==  overflow in your program's main thread (unlikely but
==12408==  possible), you can try to increase the size of the
==12408==  main thread stack using the --main-stacksize= flag.
==12408==  The main thread stack size used in this run was 8388608.
==12408==
```

### PR DESCRIPTION
**Pull request by zporky**
declaration is a CapturedDecl.

Fixes #415

### CODE REVIEW

File: plugins/cpp/parser/src/symbolhelper.cpp [hunk 1]
```diff
@@ -69,12 +74,25 @@ std::string getMangledName(
...
+      // there are contexts which getParentFunctionOrMethod() is not null
+      // but the result is not FunctionDecl
+      // like CapturedDecl	    
+      if ( 0 == llvm::dyn_cast<clang::FunctionDecl>(dc) )
+      {
+        LOG(debug) << "Yet another getparentFunctionMethod() && "
+		      "! llvm::dyn_cast<clang::FunctionDecl>(dc) ";
+	LOG(debug) << dc->getDeclKindName();
+
+	dc = dc->getLookupParent();
```
Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:93]: Formatting issue.

File: plugins/cpp/parser/src/symbolhelper.cpp [hunk 2]
```diff
@@ -69,12 +74,25 @@ std::string getMangledName(
...
   if (const clang::VarDecl* vd = llvm::dyn_cast<clang::VarDecl>(nd_))
   {
     std::string result = nd_->getQualifiedNameAsString();
 
     if (const clang::DeclContext* dc = nd_->getParentFunctionOrMethod())
     {
+      // there are contexts which getParentFunctionOrMethod() is not null
+      // but the result is not FunctionDecl
+      // like CapturedDecl	    
+      if ( 0 == llvm::dyn_cast<clang::FunctionDecl>(dc) )
```
Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:87]: Please don't write yoda code, such a coding structure isn't used in the project.

Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:87]: In addition, `dyn_cast` returns a pointer, so the convention is using `if (const auto* P = dyn_cast<...>(...))`.

If you don't need the returned cast pointer, you could also write:

`if (!llvm::isa<T>(obj))`

which returns a `bool`.

File: plugins/cpp/parser/src/symbolhelper.cpp (hunk 1 above)
Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:93]: Is this working in generic cases, or just this in particular? Is there a chance that there are *multiple* `CapturedStmt`s embedded into each other if multiple OMP pragmas are nested?

File: plugins/cpp/parser/src/symbolhelper.cpp [hunk 3]
```diff
@@ -6,9 +6,11 @@
 
 #include <model/fileloc.h>
 #include <model/fileloc-odb.hxx>
+#include <util/logutil.h>
 
 #include "symbolhelper.h"
 
+
```
Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:13]: Unnecessary change here.

File: plugins/cpp/parser/src/symbolhelper.cpp [hunk 4]
```diff
@@ -69,15 +71,31 @@ std::string getMangledName(
   const clang::NamedDecl* nd_,
   const model::FileLoc& fileLoc_)
 {
+
```
Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:74]: Unnecessary change here.

File: plugins/cpp/parser/src/symbolhelper.cpp [hunk 5]
```diff
@@ -69,15 +71,31 @@ std::string getMangledName(
...
   if (const clang::VarDecl* vd = llvm::dyn_cast<clang::VarDecl>(nd_))
   {
     std::string result = nd_->getQualifiedNameAsString();
 
     if (const clang::DeclContext* dc = nd_->getParentFunctionOrMethod())
     {
-      result += ':' + getMangledName(
-        mangleContext_,
-        llvm::dyn_cast<clang::FunctionDecl>(dc));
+      // Workaround ISSUE #415 Parser segfaults.
```
Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:81]: Please directly link the issue with an URL.

File: plugins/cpp/parser/src/symbolhelper.cpp [hunk 6]
```diff
@@ -69,15 +71,31 @@ std::string getMangledName(
...
+      // There are contexts which getParentFunctionOrMethod() is not null
+      // but the result is not FunctionDecl (like CapturedDecl)
+      // In that case we go up until we reach a Function or TU decl context
```
Comment (whisperity) [inline plugins/cpp/parser/src/symbolhelper.cpp:84]:
```suggestion
      // There are contexts in which getParentFunctionOrMethod() is not
      // null, but the result is not FunctionDecl (such as CapturedDecl),
      // e.g. when the to-be-mangled node is inside an OMP block.
      // In that case, try to find some parent DeclContext we can
      // reasonably mangle.
```

**Review by mcserep — APPROVED** [review]
Fixes #415 .

**Pull request merged**